.github/workflows/ci-cd.yml

```diff
@@ -1,54 +1,78 @@
+
 
 jobs:
-  Unit-Testing:
-    # Set up operating system
-    runs-on: ubuntu-latest
+  ################################################################################################
+  # Unit-Tests: Run unit tests using pytest
+  ################################################################################################
+  Unit-Tests:
+    runs-on: ${{ matrix.os }}
+    timeout-minutes: 15 # Consider increasing timeout
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest, macos-14] # , windows-latest removed for now
+        env: ["py310", "py311", "py312"]
 
-    # Define job steps
     steps:
-      - name: Set up Python
-        uses: actions/setup-python@v4
-        with:
-          python-version: "3.9"
-
-      - name: Check-out repository
-        uses: actions/checkout@v3
+      - uses: actions/checkout@v4
 
-      - name: Install poetry
-        uses: snok/install-poetry@v1
-
-      - name: Install package
-        run: poetry install
+      - name: Install Pixi
+        uses: prefix-dev/setup-pixi@v0.8.1
+        with:
+          environments: ${{ matrix.env }}
+          pixi-version: v0.33.0
+          cache: true
+          locked: false # wont be the same because of the tag
 
-      - name: Test with pytest
-        # Github action runners now have 4 cores
-        run: poetry run pytest -s -v -n 4 tests/
+      - name: Run pytest
+        run: |
+          pixi run -e ${{ matrix.env }} test
 
-      # Commented out until we have a repo for coverage
-      # - name: Test with pytest
-      #   run: poetry run pytest tests/ --cov=readii --cov-report=xml
+      - name: Upload coverage report artifact to be used by Codecov
+        # only upload if matrix.os is ubuntu-latest and matrix.python-version is 3.12
+        if: matrix.os == 'ubuntu-latest' && matrix.env == 'py312'
+        uses: actions/upload-artifact@v4
+        with:
+          name: coverage-report
+          path: coverage-report
+
+################################################################################################
+  # Semantic-Release: Run semantic-release to automate versioning and publishing
+  ################################################################################################
+  Semantic-Release:
+    permissions:
+      # Read releases read-all
+      contents: write
+      packages: write
+      issues: write
+      pull-requests: write
+
+    needs: [Unit-Tests]
+
+    # if pulling to main, deploy to PyPI
+    if: github.ref == 'refs/heads/main'
 
-      # - name: Use Codecov to track coverage
-      #   uses: codecov/codecov-action@v3
-      #   with:
-      #     files: ./coverage.xml   # coverage report
+    env:
+      PIXI_ENV: "publish"
 
-  Continuous-Deployment:
-    needs: Unit-Testing
-
-    # # Only run this job if new work is pushed to "main"
-    if: github.ref == 'refs/heads/main'
+    strategy:
+      matrix:
+        os: [ubuntu-latest]
+        python-version: ["3.12"]
```

Comment:
Has this worked? I feel like a lot of packages I've used were with Python 3.9 so I wonder if tests are passing?

Reply:
I've extended the testing suite to run all the pytests on python `3.10`, `3.11` and `3.12`
combined with GHA matrix expansion across different machines, all the tests seem to pass. 
![image](https://github.com/user-attachments/assets/271ac59b-4e33-4451-94c6-d95bff8e8e4a)

The `release` step hasnt been tested yet due to not being on `main` though. 

note: cant test yet for `3.13` due to [a bug in scikit image](https://github.com/scikit-image/scikit-image/pull/7448)